packages/mui-base/src/useNumberInput/useNumberInput.ts

```diff
@@ -0,0 +1,371 @@
+import * as React from 'react';
+import MuiError from '@mui/utils/macros/MuiError.macro';
+import {
+  // unstable_useControlled as useControlled, // TODO: do I need this?
+  unstable_useForkRef as useForkRef,
+} from '@mui/utils';
+import { FormControlUnstyledState, useFormControlUnstyledContext } from '../FormControlUnstyled';
+import {
+  UseNumberInputParameters,
+  UseNumberInputRootSlotProps,
+  UseNumberInputInputSlotProps,
+  UseNumberInputIncrementButtonSlotProps,
+  UseNumberInputDecrementButtonSlotProps,
+  UseNumberInputReturnValue,
+} from './useNumberInput.types';
+import { clamp, isNumber } from './utils';
+import extractEventHandlers from '../utils/extractEventHandlers';
+
+type StepDirection = 'up' | 'down';
+
+// TODO
+// 1 - make a proper parser
+// 2 - accept a parser (func) prop
+const parseInput = (v: string): string => {
+  return v ? String(v.trim()) : String(v);
+};
+
+/**
+ *
+ * Demos:
+ *
+ * - [Unstyled Number Input](https://mui.com/base/react-number-input/#hook)
+ *
+ * API:
+ *
+ * - [useNumberInput API](https://mui.com/base/react-number-input/hooks-api/#use-number-input)
+ */
+export default function useNumberInput(
+  parameters: UseNumberInputParameters,
+): UseNumberInputReturnValue {
+  const {
+    min,
+    max,
+    step,
+    shiftMultiplier = 10,
+    defaultValue: defaultValueProp,
+    disabled: disabledProp = false,
+    error: errorProp = false,
+    onBlur,
+    onChange,
+    onFocus,
+    onValueChange,
+    required: requiredProp = false,
+    value: valueProp,
+    inputRef: inputRefProp,
+  } = parameters;
+
+  // TODO: make it work with FormControl
+  const formControlContext: FormControlUnstyledState | undefined = useFormControlUnstyledContext();
+
+  const { current: isControlled } = React.useRef(valueProp != null);
+
+  const handleInputRefWarning = React.useCallback((instance: HTMLElement) => {
+    if (process.env.NODE_ENV !== 'production') {
+      if (instance && instance.nodeName !== 'INPUT' && !instance.focus) {
+        console.error(
+          [
+            'MUI: You have provided a `slots.input` to the input component',
+            'that does not correctly handle the `ref` prop.',
+            'Make sure the `ref` prop is called with a HTMLInputElement.',
+          ].join('\n'),
+        );
+      }
+    }
+  }, []);
+
+  const inputRef = React.useRef<HTMLInputElement>(null);
+  const handleInputRef = useForkRef(inputRef, inputRefProp, handleInputRefWarning);
+
+  const [focused, setFocused] = React.useState(false);
+
+  // the "final" value
+  const [value, setValue] = React.useState(valueProp ?? defaultValueProp);
+  // the (potentially) dirty or invalid input value
+  const [dirtyValue, setInputValue] = React.useState<string | undefined>(undefined);
```

Comment:
Let's make sure the names match:

```suggestion
  const [dirtyValue, setDirtyValue] = React.useState<string | undefined>(undefined);
```